api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java

```diff
@@ -102,7 +102,15 @@ private boolean eval(ContentFile<?> file) {
       this.lowerBounds = file.lowerBounds();
       this.upperBounds = file.upperBounds();
 
-      return ExpressionVisitors.visitEvaluator(expr, this);
+      try {
+        return ExpressionVisitors.visitEvaluator(expr, this);
+      } finally {
+        // allow temporary state to be collected because this is in a thread-local
+        this.valueCounts = null;
```

Comment:
I think this is subtle and hard to understand on why this is needed. I prefer the the new visitor approach as you guys discussed in #1139.

Reply:
That sounds good to me.

Reply:
Thanks @rdsr and @rdblue for the comments. Updated the PR accordingly to remove the thread local objects and use new visitors to keep the code straightforward.